internal/driver/driver.go

```diff
@@ -0,0 +1,793 @@
+	s.lc.Debugf("Device %s is updated", deviceName)
+	return nil
+}
+
+// RemoveDevice is a callback function that is invoked
+// when a Device associated with this Device Service is removed
+func (s *Driver) RemoveDevice(deviceName string, protocols map[string]models.ProtocolProperties) error {
+	s.lc.Debugf("Device %s is removed", deviceName)
+	return nil
+}
+
+// Discover triggers protocol specific device discovery, which is an asynchronous operation.
+// Devices found as part of this discovery operation are written to the channel devices.
+// func (s *Driver) Discover() {
+
+// }
+
+func (s *Driver) ValidateDevice(device models.Device) error {
+
+	return nil
+}
+
+// Create S7Client by 'Device' definition
+func (s *Driver) NewS7Client(deviceName string, protocol map[string]models.ProtocolProperties) *S7Client {
+	// protocol := device.Protocols[Protocol]
+	// fmt.Println(protocol)
+	pp := protocol[Protocol]
+	var errt error
+	host, errt := cast.ToStringE(pp["Host"])
+	if errt != nil {
+		s.lc.Errorf("Timeout not found or not an integer in Protocol, USE DEFAULT 30s error: %s", errt)
+		return nil
+	}
+	port, errt := cast.ToStringE(pp["Port"])
+	if errt != nil {
+		s.lc.Errorf("Timeout not found or not an integer in Protocol, USE DEFAULT 30s error: %s", errt)
+		return nil
+	}
+	rack, errt := cast.ToIntE(pp["Rack"])
+	if errt != nil {
+		s.lc.Errorf("Timeout not found or not an integer in Protocol, USE DEFAULT 30s error: %s", errt)
+		return nil
+	}
+	slot, errt := cast.ToIntE(pp["Slot"])
+	if errt != nil {
+		s.lc.Errorf("Timeout not found or not an integer in Protocol, USE DEFAULT 30s error: %s", errt)
+		return nil
+	}
+	timeout, errt := cast.ToIntE(pp["Timeout"])
+	if errt != nil {
+		s.lc.Errorf("Timeout not found or not an integer in Protocol, USE DEFAULT 30s error: %s", errt)
+		timeout = 30
+	}
+	idletimeout, errt := cast.ToIntE(pp["IdleTimeout"])
+	if errt != nil {
+		s.lc.Errorf("IdleTimeout not found or not an ingeger in Protocol, USE DEFAULT 30s, error: %s", errt)
+		idletimeout = 30
+	}
+
+	// create handler: PLC tcp client
+	// fmt.Println(host, rack, slot)
+	handler := gos7.NewTCPClientHandler(host+":"+port, rack, slot)
+	if handler == nil {
+		s.lc.Errorf("Cant not create NewTCPClientHandler: %s", handler)
+		return nil
+	}
+	s.lc.Infof("New TCP Client: %s", handler)
+
+	// handler connect timeout from 'Timeout'
+	handler.Timeout = time.Duration(timeout) * time.Second
+
+	// handler connect idle timeout from 'IdleTimeout'
+	handler.IdleTimeout = time.Duration(idletimeout) * time.Second
+
+	// connect to S7
+	err := handler.Connect()
+	if err != nil {
+		s.lc.Errorf("Can't handler S7 Connect: %s, error: %s", deviceName, err)
+		// return nil
+	}
+
+	s7client := gos7.NewClient(handler)
+	client := &S7Client{
+		DeviceName: deviceName,
+		Client:     s7client,
+	}
+	return client
+
+}
+
+// Get S7Client by 'DeviceName'
+func (s *Driver) getS7Client(deviceName string) *S7Client {
+	s7Client := s.s7Clients[deviceName]
```

Comment:
Add locking around access to `s7Clients`

Reply:
add locking.